### PR TITLE
[Shop][Product] Viewing product's price

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | -
| License         | MIT

### CODE REVIEW

File: features/product/viewing_products/viewing_products_price.feature [hunk 1]
```diff
@@ -0,0 +1,14 @@
+@viewing_products
+Feature: Viewing a product price
+  In order to see products price
```
Comment (pamil) [inline features/product/viewing_products/viewing_products_price.feature]: Aren't we using 4 spaces as indentation in other features?

File: features/product/viewing_products/viewing_product_price_on_products_list.feature [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+    Background:
+        Given the store operates on a single channel in "France"
+
+    @todo
+    Scenario: Viewing a products with price on list
+        Given the store has a product "T-shirt watermelon" priced at "€19.00"
+        And this product belongs to "T-Shirt"
+        When I browse products from taxon "T-Shirts"
+        Then I should see the product "T-shirt banana" with price "€19.00"
```
Comment (NeverResponse) [inline features/product/viewing_products/viewing_product_price_on_products_list.feature]: I might not be a professional biologist, but i'm quite certain that bananas are not watermelons.

File: features/product/viewing_products/viewing_product_price_on_products_list.feature [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    Background:
+        Given the store operates on a single channel in "France"
+
+    @ui
+    Scenario: Viewing a products with price on list
+        Given the store has a product "T-shirt watermelon" priced at "€19.00"
+        And the store classifies its products as "T-Shirts"
+        And this product belongs to "T-Shirts"
+        When I browse products from taxon "T-Shirts"
+        Then I should see the product "T-shirt banana" with price "€19.00"
```
Comment (lchrusciel) [inline features/product/viewing_products/viewing_product_price_on_products_list.feature]: Not really

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 4]
```diff
@@ -163,4 +164,27 @@ public function iShouldSeeEmptyListOfProducts()
...
+            'Product should have price %2$s, but it has %s'
+        );
+    }
+
+    /**
+     * @Then I should see the product :productName with price :productPrice
+     */
+    public function iShouldSeeTheProductWithPrice($productName, $productPrice)
+    {
+        Assert::false(
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: true

File: src/Sylius/Component/Variation/spec/Resolver/DefaultVariantResolverSpec.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        $this->shouldImplement(VariantResolverInterface::class);
+    }
+
+    function it_returns_first_variant(
+        VariableInterface $product,
+        VariantInterface $variant,
+        Collection $variants
+    ) {
+
```
Comment (lchrusciel) [inline src/Sylius/Component/Variation/spec/Resolver/DefaultVariantResolverSpec.php]: Redundant

File: src/Sylius/Component/Variation/Resolver/DefaultVariantResolver.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+namespace Sylius\Component\Variation\Resolver;
+
+use Sylius\Component\Variation\Model\VariableInterface;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class DefaultVariantResolver implements VariantResolverInterface
```
Comment (lchrusciel) [inline src/Sylius/Component/Variation/Resolver/DefaultVariantResolver.php]: final

File: src/Sylius/Bundle/CoreBundle/Cart/ItemResolver.php [hunk 7]
```diff
@@ -60,27 +61,35 @@ class ItemResolver implements ItemResolverInterface
...
-        AvailabilityCheckerInterface   $availabilityChecker,
-        DelegatingCalculatorInterface  $priceCalculator,
-        ChannelContextInterface        $channelContext
+        CartProviderInterface $cartProvider,
+        RepositoryInterface $productRepository,
+        FormFactoryInterface $formFactory,
+        AvailabilityCheckerInterface $availabilityChecker,
+        DelegatingCalculatorInterface $priceCalculator,
+        ChannelContextInterface $channelContext,
+        VariantResolverInterface $defaultVariantResolver
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Cart/ItemResolver.php]: 7 injected services? We should consider some refactoring and extract some of this logic... It starts to be too complicated.

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 8]
```diff
@@ -407,9 +417,11 @@ private function createProduct($productName, $price = 0)
         $product = $this->productFactory->createWithVariant();
 
         $product->setName($productName);
-        $product->getFirstVariant()->setPrice($price);
+        $variant = $this->defaultVariantResolver->getVariant($product);
+        $variant->setPrice($price);
         $product->setCode($this->convertToCode($productName));
-        $product->getFirstVariant()->setCode($product->getCode());
+        $variant = $this->defaultVariantResolver->getVariant($product);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: Redundant

File: features/product/viewing_products/viewing_product_price_on_products_list.feature [hunk 9]
```diff
@@ -0,0 +1,16 @@
+@viewing_products
+Feature: Viewing a product price on products list
+    In order to see products price
```
Comment (NeverResponse) [inline features/product/viewing_products/viewing_product_price_on_products_list.feature]: `In order to see the prices of listed products`?

File: features/product/viewing_products/viewing_products_price.feature [hunk 10]
```diff
@@ -0,0 +1,14 @@
+@viewing_products
+Feature: Viewing a product price
```
Comment (lchrusciel) [inline features/product/viewing_products/viewing_products_price.feature]: Shouldn't this na above example be coupled?

File: src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/ProductOptionFieldSubscriber.php [hunk 11]
```diff
@@ -23,6 +24,19 @@
 class ProductOptionFieldSubscriber implements EventSubscriberInterface
 {
     /**
+     * @var VariantResolverInterface
+     */
+    private $defaultVariantResolver;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/ProductOptionFieldSubscriber.php]: It does not have to be a default one. Lets change `defaultVariantResolver` to `variantResolver` here, and in all other occurrences.

File: src/Sylius/Bundle/CoreBundle/Controller/ProductController.php [hunk 12]
```diff
@@ -271,13 +271,16 @@ public function searchAction(Request $request)
         $results = [];
         $products = $this->container->get('sylius.repository.product')->createFilterPaginator($request->query->get('criteria'));
         $helper = $this->container->get('sylius.templating.helper.currency');
+        $defaultVariantResolver = $this->container->get('sylius.resolver.default_variant');
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductController.php]: `sylius.resolver.default_variant` -> `sylius.variant_resolver.default`?

Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductController.php]: Won't be needed after https://github.com/Sylius/Sylius/pull/5643 :tada:

Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Controller/ProductController.php]: In here, yes. But in general, I would suggest to change name of this service from `sylius.resolver.default_variant` to `sylius.variant_resolver.default`

File: src/Sylius/Bundle/CoreBundle/Cart/ItemResolver.php (hunk 7 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Cart/ItemResolver.php]: @lchrusciel no worries, this class will probably be removed soon 😉

File: src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/ProductOptionFieldSubscriber.php [hunk 13]
```diff
@@ -45,7 +59,7 @@ public function preSetData(FormEvent $event)
         $form = $event->getForm();
 
         /** Options should be disabled for configurable product if it has at least one defined variant */
-        $disableOptions = (null !== $product->getFirstVariant()) && (false === $product->hasVariants());
+        $disableOptions = (null !== $this->defaultVariantResolver->getVariant($product)) && (false === $product->hasVariants());
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ProductBundle/Form/EventSubscriber/ProductOptionFieldSubscriber.php]: Actually current logic can be simplified to `$disableOptions = $product->hasVariants()`

File: features/product/viewing_products/viewing_products_price.feature (hunk 10 above)
Comment (tuka217) [inline features/product/viewing_products/viewing_products_price.feature]: I'm not sure. It is almost the same, but first action is on taxon list and the second one is on product page.

**Comment by pjedrzejewski** [comment]
Nice work Ania!

**Pull request merged**